### DISCUSSION TITLE
How can I use the E2SProtocol or S2EProtocol

### DISCUSSION THREAD

**Discussion by AFLuo** [tuneinsight/lattigo#421] [Q&A]
I'm now developing an application for research purpose and I need to convert from the dbgv to a multiparty setting, so i wonder how can i use the E2SProtocol or S2EProtocol for my purpose. I don't see any examples of these two protocols. Can you guys give an example or simply illustrate the overall procedure?

**Comment by Pro7ech** [comment]
Hi @AFLuo, you can take a look at the associated test:

https://github.com/tuneinsight/lattigo/blob/1f00f3ca7147a13f1313c64610385a904f84db7e/mhe/mheint/mheint_test.go#L163

**Reply by AFLuo** [reply]
I'm currently working with lattigo v4 since I need the `bgv.EncodeCoeffs` to implement similar encoding in this paper: [Cheetah](https://www.usenix.org/system/files/sec22-huang-zhicong.pdf). I notice that v5 has not yet included the coefficient-wise encoding so i need to stay v4. I wrote the following code to mimic the test code:
```
func TestE2SProtocol(params bgv.Parameters,
	nParties int,
	ciphertext *rlwe.Ciphertext) {

	type Party struct {
		e2s         *dbgv.E2SProtocol
		s2e         *dbgv.S2EProtocol
		sk          *rlwe.SecretKey
		publicShare *drlwe.CKSShare
		secretShare *rlwe.AdditiveShare
	}

	P := make([]*Party, nParties)
	for i := range P {
		if i == 0 {
			P[i].e2s = dbgv.NewE2SProtocol(params, 3.2)
			P[i].s2e = dbgv.NewS2EProtocol(params, 3.2)
		} else {
			P[i].e2s = P[0].e2s.ShallowCopy()
			P[i].s2e = P[0].s2e.ShallowCopy()
		}
		P[i].sk = bgv.NewKeyGenerator(params).GenSecretKey()
		P[i].publicShare = P[i].e2s.AllocateShare(params.MaxLevel())
		// how to convert a bgv parameter to a rlwe parameter
		P[i].secretShare = rlwe.NewAdditiveShare(params)
	}
	for i, p := range P {
		p.e2s.GenShare(p.sk, ciphertext, p.secretShare, p.publicShare)
		if i > 0 {
			p.e2s.AggregateShares(P[0].publicShare, p.publicShare, P[0].publicShare)
		}
	}
	P[0].e2s.GetShare(P[0].secretShare, P[0].publicShare, ciphertext, P[0].secretShare)

}
```
So i wonder if the usage of functions is correct. Besides, i wonder how to extract the coeffs in each P[i]'s secret share. It would be better if i can set the value by myself. By the way, the line `P[i].secretShare = rlwe.NewAdditiveShare(params)` actually gives a error message say that `rlwe.NewAdditiveShare` takes in only `rlwe.Parameters` instead of `bgv.Parameters` so i wonder if there is any method to convert a `bgv.Parameters` to `rlwe.Parameters`.
Really appreciate your efforts. Looking forward to your reply.

**Comment by Pro7ech** [comment]
Hi @AFLuo, could you provide more context around the applications that you are trying to build? This might help me guide you toward what you are trying to achieve.

By the way, Lattigo v5 does implement coefficient encoding. This is now done through setting the flag `plaintext.IsBatched=false` before encoding/decoding.

**Reply by AFLuo** [reply]
Thank you for your patience @Pro7ech.  I'm now trying to build a privacy-preserving solution to enable joint data analysis across different Data Providers (DPs). My solution employs both MHE and MPC. For HE context, I adopt the idea from [Cheetah](https://www.usenix.org/system/files/sec22-huang-zhicong.pdf) to enable inner products over HE ciphertexts without any rotations. I have implemented [Cheetah](https://www.usenix.org/system/files/sec22-huang-zhicong.pdf) by carefully re-arranging the integer slice before feeding it to the coefficient-wise encoding.  So here's what I need

1. Coefficient-wise encoding;
2. Flexible conversion between MHE and MPC.

I have already implemented [Cheetah](https://www.usenix.org/system/files/sec22-huang-zhicong.pdf) in lattigo v4 in the context of BGV/DBGV. So I'm now trying to implement conversions between MHE and MPC. Specifically, I adopt n-out-of-n additive secret sharing, same as [semi2k](https://github.com/data61/MP-SPDZ/) from MP-SPDZ but I don't rely on their framework.  So i'm now stuck by implementing the conversions. To implement it, I want to know how to convert the secret share in the context of lattigo to the context of general ring/field element represented as an `uint` number.
